File: backend/app/schemas/tag.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# REMOVED: FieldSchemaResponse import - no longer needed since we don't include nested schema


class TagBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="Tag name")
    color: str | None = Field(
        None, pattern=r"^#[0-9A-Fa-f]{6}$", description="Hex color code"
    )


class TagCreate(TagBase):
    schema_id: UUID | None = None  # BUG FIX: Allow schema binding during creation
    is_video_type: bool = (
        True  # True = category (one per video), False = label (multiple)
    )


class TagUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    color: str | None = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")
    schema_id: UUID | None = None  # REF MCP: Simplified from Field(default=None)
    is_video_type: bool | None = None  # Optional: change category/label type


class TagResponse(TagBase):
    id: UUID
    user_id: UUID
    schema_id: UUID | None = None  # Schema FK (use this instead of nested object)
    is_video_type: bool = True  # Default True for legacy tags without this field
    # REMOVED: schema: FieldSchemaResponse | None - Causes lazy-loading issues
    # Frontend only needs schema_id - if full schema needed, fetch via GET /api/schemas/{id}
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/tag.ts
import { z } from "zod";

/**
 * Zod schema for Tag validation
 * Matches backend API response (UUID and datetime as strings after JSON serialization)
 */
export const TagSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1).max(100),
  color: z
    .string()
    .regex(/^#[0-9A-Fa-f]{6}$/)
    .nullable(),
  schema_id: z.string().uuid().nullish(), // null/undefined = no schema, UUID = bound to schema
  is_video_type: z.boolean(), // true = video type/category, false = label
  user_id: z.string().uuid(),
  created_at: z.string(),
  updated_at: z.string(),
});

/**
 * Tag type inferred from Zod schema
 */
export type Tag = z.infer<typeof TagSchema>;

/**
 * Schema for creating a new tag
 */
export const TagCreateSchema = z.object({
  name: z.string().min(1).max(100),
  color: z
    .string()
    .regex(/^#[0-9A-Fa-f]{6}$/)
    .optional(),
  schema_id: z.string().uuid().nullish(), // nullish = nullable + optional (Zod v3.23+)
  is_video_type: z.boolean().optional(), // defaults to true server-side
});

export type TagCreate = z.infer<typeof TagCreateSchema>;

/**
 * Schema for updating a tag
 */
export const TagUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  color: z
    .string()
    .regex(/^#[0-9A-Fa-f]{6}$/)
    .nullable()
    .optional(),
  schema_id: z.string().uuid().nullish(), // Can update or remove schema binding
  is_video_type: z.boolean().optional(), // Can change tag type
});

export type TagUpdate = z.infer<typeof TagUpdateSchema>;

/**
 * Array schema for validating multiple tags
 */
export const TagsSchema = z.array(TagSchema);
